icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -257,16 +258,37 @@ int32_t FormattedStringBuilder::prepareForInsert(int32_t index, int32_t count, U
     U_ASSERT(count >= 0);
     if (index == 0 && fZero - count >= 0) {
         // Append to start
+        int32_t newLength;
+        // Add and return U_INPUT_TOO_LONG_ERROR if overflow int32_t.
+        if (uprv_add32_overflow(fLength, count, &newLength)) {
+          status = U_INPUT_TOO_LONG_ERROR;
+          return count;
+        }
         fZero -= count;
-        fLength += count;
+        fLength = newLength;
         return fZero;
-    } else if (index == fLength && fZero + fLength + count < getCapacity()) {
-        // Append to end
-        fLength += count;
-        return fZero + fLength - count;
```

Comment:
I would preserve this append-to-end optimization as well, by refactoring the test slightly to simply avoid the addition:
```
    } else if (index == fLength && count <= getCapacity() - (fZero + fLength)) {
        // Append to end
        fLength += count;
        return fZero + fLength - count;
```

Only the need-to-grow-the-buffer "else" then needs more work.

Reply:
how could we be sure that (fZero + fLength) does not overflow int32_t here?

Reply:
done